scripts/src/public/services/orders.ts

```diff
@@ -151,65 +143,100 @@ export async function createMarketplaceOrder(offerId: string, user: User, logger
+		status: "opened",
+		blockchainData: {
+			sender_address: app.walletAddresses.sender,
+			recipient_address: recipient.walletAddress
+		}
+	}, {
+		type: "earn",
+		user: recipient,
+		userId: recipient.id,
+		meta: pick(jwt.recipient, "title", "description")
+	});
+}
+
+async function createNormalSpendExternalOrder(sender: User, jwt: ExternalSpendOrderJWT) {
+	const app = await Application.findOneById(sender.appId);
+
+	if (!app) {
+		throw NoSuchApp(sender.appId);
+	}
+
+	return db.ExternalOrder.new({
+		offerId: jwt.offer.id,
+		amount: jwt.offer.amount,
+		status: "opened",
+		blockchainData: {
+			sender_address: sender.walletAddress,
+			recipient_address: app.walletAddresses.recipient
+		}
+	}, {
+		type: "spend",
+		user: sender,
+		userId: sender.id,
+		meta: pick(jwt.sender, "title", "description")
+	});
+}
+
 export async function createExternalOrder(jwt: string, user: User, logger: LoggerInstance): Promise<OpenOrder> {
 	const payload = await validateExternalOrderJWT(jwt, user.appUserId, logger);
 
 	let order = await db.Order.findBy(payload.offer.id, user.id);
 
 	if (!order || order.status === "failed") {
-		const app = await Application.findOneById(user.appId);
-		if (!app) {
-			throw NoSuchApp(user.appId);
-		}
-
-		let title: string;
-		let description: string;
-		let sender_address: string;
-		let recipient_address: string;
-
-		if (payload.sub === "earn") {
-			title = (payload as ExternalEarnOrderJWT).recipient.title;
-			description = (payload as ExternalEarnOrderJWT).recipient.description;
-			sender_address = app.walletAddresses.sender;
-			recipient_address = user.walletAddress;
+		if (isPayToUser(payload)) {
+			order = await createP2PExternalOrder(user, payload);
+		} else if (isExternalEarn(payload)) {
+			order = await createNormalEarnExternalOrder(user, payload);
 		} else {
-			// spend or pay_to_user
-			await addWatcherEndpoint([app.walletAddresses.recipient]);  // XXX how can we avoid this and only do this for the first ever time we see this address?
-			title = (payload as ExternalSpendOrderJWT).sender.title;
-			description = (payload as ExternalSpendOrderJWT).sender.description;
-			sender_address = user.walletAddress;
-			// TODO in case of pay_to_user, needs another lookup for the recipient_user_wallet
-			recipient_address = app.walletAddresses.recipient;
+			order = await createNormalSpendExternalOrder(user, payload);
 		}
 
-		order = db.ExternalOrder.new({
-			offerId: payload.offer.id,
-			amount: payload.offer.amount,
-			type: payload.sub,
-			status: "opened",
-			blockchainData: {
-				sender_address,
-				recipient_address
-			}
-		});
 		await order.save();
 
-		const context = db.OrderContext.new({
-			user, order,
-			orderId: order.id,
-			userId: user.id,
-			meta: {
-				title,
-				description
-			},
-			role: payload.sub === "spend" ? "sender" : "recipient"
+		order.contexts.forEach(context => {
+			metrics.createOrder("external", context.type, payload.offer.id);
```

Comment:
I'd only report this once and change the type to p2p here

Reply:
what about the type?